Masked frames and SkyCoord

mhvk opened the pull request:
This pull request is the follow-up to #16845, using the mask on representations in `SkyCoord` and coordinate frames. It also includes brief docs and a what's-new entry.

Notes:
- Based on #16845 ~(hence opened as draft for now)~, but feel free to just look at the whole here;
- Because the coordinates API is shown without inherited members, I cannot link to the new `.mask` property and other features in the what's-new. I tried just adding  `:inherited-members:` to `ref_api.py`, but that leads to large numbers of failures that seem out of scope of this PR.
- So far, I've stuck to the standard `Masked` approach, that masks just get propagated. For some methods, like `apply_space_motion`, as well as for coordinate transformations, it might make more sense to treat masked values truly as missing ones, so that, e.g., a masked distance is taken to be the same as if the coordinate had `UnitSpherical` representation. The question then is whether that is in fact what the user wants, i.e., to what extent one should refuse the temptation to guess...

<!-- Optional opt-out -->

- [X] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

mhvk commented:
I'm taking this out of draft given the upcoming feature-freeze deadline -- if we miss that, masked `SkyCoord` will not be possible until astropy 8.0. 

I think this is complete if somewhat minimal, with the main question whether one should go beyond simply propagating masks (as done here). E.g., what should expected behaviour be for coordinate conversions and methods like `apply_space_motion` for spherical representations with some masked distances, PMs, and radial velocities (see also on top).

My own 2¢ is that we should simply go with the minimal implementation I have, mark coordinate masks as experimental, and allow masks not to propagate in some specific cases later.

cc @taldcroft, @StuartLittlefair, @adrn, @eerovaher, @nstarman

nstarman commented:
I like the look of this a lot, but I'm not sure about the special treatment of spherical coordinates to disregard the distance. A few things: 1. If a user passes a masked distance and unmasked angles I think they would be very surprised to not find that in the total mask. 2. I think if a user wants to mark a distance as truly missing they should use Nan. Which of course can be combined with a mask, but these have different meanings, right?

mhvk commented:
> I like the look of this a lot, but I'm not sure about the special treatment of spherical coordinates to disregard the distance.

Thanks for bringing that up! It is the thing I have most struggled with, what exactly to do with missing distances, as well as incomplete differentials (proper motions and radial velocities). The basic problem is that a `SkyCoord` often just has RA, Dec, and for many catalogues there will be more informationfor only a subset of the data. But RA, Dec is enough to know where to point one's telescope!

For the mask, I hope my `.get_mask()` function provides the flexibility to get what exactly a user wants, but it is then a question of defaults, with two most logical options:
1. `.mask` tells whether "one can point a telescope" and default `.get_mask()` combines the masks of everything.
2. `.mask` combines everything, while `.get_mask("ra", "dec")` (perhaps the default) just does RA, Dec.

Which one would you prefer?

p.s. Part of the reason to have masks is to allow table joins in which one table does not have the given `SkyCoord` column. For that, of course, just any mask will do.

mhvk commented:
> Have you thought about how `SkyCoord.apply_space_motion()` should work with masked time? I don't see any tests for that.

If I'm right, it will "just work", i.e., propagate, but let's add a test!

mhvk commented:
Test added - it works as expected.

nstarman commented:
> Just to be sure: still exclude in the mask possible attributes, right, i.e., effectively just return self.data.mask (would be consistent with Time).

That's an interesting Q. Have to think on that more. I would say that if we're marking this as an experimental feature then yes, for now. We can revisit the question of attributes after v7.

> It might then make more sense to simply add an argument to the method about how to deal with masked data (not sure what is best, but we would now have the benefit of not having to decide too much...).

Agreed, if I'm interpreting this right and you mean figuring out the added arguments can happen post v7.
The alternative to added args is a context manager approach, which might also be good

```python
with mask_propagation_controller(...):
    ...
```

But can we decide this for v7.1?

eerovaher reviewed (commented):
>> I lean towards the 2nd one. In the lower classes .mask combines all masks. IMO we should keep that pattern and make sure that other common use cases are deeply supported.

> Just to be sure: still exclude in the mask possible attributes, right, i.e., effectively just return self.data.mask (would be consistent with Time).

Consistency is good!

mhvk commented on astropy/coordinates/baseframe.py:
```diff
@@ -359,6 +361,64 @@ def _insert(self, obj, values, axis=0):
         return out
 
 
+class CoordinateSharedMethods:
```
The problem is that while just letting `__getattr__` take care works for `masked` and `mask`, it does not work for `.get_mask`, since it takes a series of attributes of the class and not all `SkyCoord` attributes are available on the frame (this is explicitly tested with `pressure`), i.e., `self` is required to the `SkyCoord`, not its underlying frame.

p.s. Of course, having this shared method mixin will also help remove duplication between the two like for `.insert()`.

mhvk replied:
p.p.s. A side benefit of having a mixin class is that, eventually, we can actually have sphinx render the relevant properties and methods (once `:inherited-members:` works properly).

mhvk replied:
Yes, just to keep things together, and with a view of expanding later. But if this is something that will prevent this PR from getting into 7.0, I'm fine with removing this update, and just duplicating `SkyCoord.get_mask`...

eerovaher replied:
`CoordinateSharedMethods` looks like a private implementation detail that we should be able to change without notice, but it does implement methods that are clearly a part of the public API. I am very hesitant to introduce such classes to `coordinates` until the APE about public API has been accepted.

I would implement methods on `BaseCoordinateFrame` if `SkyCoord` can use them via `__getattr__()` because there are several methods that are implemented like that already. For `get_mask()` it might be best to avoid code duplication by introducing a private `_get_mask()` that both `SkyCoord.get_mask()` and `BaseCoordinateFrame.get_mask()` can call.

mhvk replied:
Yes, fair point that really it is best not to mix new types of implementation in. Will make the suggested changes.

mhvk commented on astropy/coordinates/baseframe.py:
```diff
@@ -359,6 +361,64 @@ def _insert(self, obj, values, axis=0):
+    def mask(self):
+        """The mask, taken from the underlying representation.
+
+        For spherical representations, the mask of the distance is ignored,
+        i.e., one gets the combined mask of the longitude and latitude.
```
OK on no special casing (see general comments). But I disagree this should be pushed down to the representations - I'd like those to be ignorant of astronomy! Though once we move the option to special-case to `get_mask()`, I can see pushing an option to just do angles down to `SpecialRepresentation.get_mask()`.

nstarman replied:
I agree. Separation of concerns + fine-grain options in get_mask.

mhvk reviewed (commented):
Thanks, @nstarman and @eerovaher. Looks like both of you are pretty clear you want the standard `.mask` just OR together all the masks in the representation and differential. That certainly is a valid choice, and simpler, so I'll go ahead and change it.  Some smaller replies inline.

mhvk commented:
OK, now implemented a default mask that combines representation and possible differentials. It makes things quite a bit simpler, so that argues that it was the right change to make.

Check PR changelog error is because this is based on the mask for representations PR, #16845, which is not yet merged.

nstarman commented on docs/coordinates/index.rst:
```diff
@@ -425,6 +425,56 @@ and transforming velocities. These are available both via the lower-level
+        [(15.00000668, 3.00000101, 0.1), (        ———,        ———, ———)]>
+
+In the last example, you will notice that the angles of the second item have
+become masked too. This is because the distance is used in the conversion
```
Not for this PR, but for frame transforms that are only rotations we could avoid going to the cartesian representation, instead doing change of basis stuff on the rotation matrix.

mhvk replied:
The problematic code is https://github.com/astropy/astropy/blob/1b82175461b87745e5328fc1622ee8e30b38fa1b/astropy/coordinates/transformations/affine.py#L118-L131

I tried a simple removal of the conversion to cartesian, since the matrix multiplication can be handled by the different representation classes internally. Unfortunately, adding the offset right now does require moving to cartesian. Obviously, this is solvable too, but best in a different PR!

mhvk replied:
Yes, sounds like a good idea to change the transformation example -- and maybe mention in the general warning about this being new that sometimes masks may propagate further than expected.

eerovaher replied:
We should open an issue so that any users who might want to report this wouldn't have to write one up themselves, but doing that can wait until the feature freeze has happened.

eerovaher commented on astropy/coordinates/tests/test_masked.py:
```diff
@@ -0,0 +1,364 @@
+        ],
+    )
+    def test_apply_space_motion(self, dt):
+        sc = self.sc.apply_space_motion(dt=dt)
```
There should be tests for unmasking the propagated coordinates.

mhvk replied:
Done

eerovaher commented on docs/whatsnew/7.0.rst:
```diff
@@ -288,6 +289,29 @@ New ``Lorentz2D`` model
+Like more generally for |Masked| instances, one can access the mask of using
+``.mask``, check whether the underlying data are masked using ``masked``,
+access underlying unmasked data using ``.unmasked``, and get a new instance
+with masked data replaced with a fill value using the ``filled()`` method.
```
Why aren't there any tests for this?

mhvk replied:
Because it just uses `__setitem__`... But fair enough, one should at least have one test!

mhvk replied:
Done

eerovaher commented on docs/whatsnew/7.0.rst:
```diff
@@ -288,6 +289,29 @@ New ``Lorentz2D`` model
+``.get_mask()`` method. For instance, ``sc.get_mask("ra", "dec")`` would
+get the combined masks of just the angular components, which may be useful
+if dealing with, e.g., a catalogue of stars where distances are available
+only for some of the objects.
```
It should also be mentioned that the frame attributes can be masked too and accessed with `get_mask()`.

mhvk replied:
Arguably too much for the what's-new, but would definitely be good to add to the actual docs.

eerovaher commented on astropy/coordinates/baseframe.py:
```diff
@@ -359,6 +361,59 @@ def _insert(self, obj, values, axis=0):
+
+        Parameters
+        ----------
+        *attrs : str
+            Attributes from which to get the masks to combine. Items can be
+            dotted, like ``"data.lon", "data.lat"``. By default, get the
+            combined mask of all components (including from differentials).
```
The docstring should mention that frame attributes can accessed too, but they are not accessed by default.

eerovaher commented on astropy/coordinates/baseframe.py:
```diff
@@ -359,6 +361,59 @@ def _insert(self, obj, values, axis=0):
+            Attributes from which to get the masks to combine. Items can be
+            dotted, like ``"data.lon", "data.lat"``. By default, get the
+            combined mask of all components (including from differentials).
+
```
Why is there no `Returns` section?

mhvk replied:
Oops...

eerovaher reviewed (commented):
Based on the tests and the documentation the public API seems mostly reasonable. The one question I have is whether the output of `get_mask()` should be a structured array so that its components could be accessed by the name of the corresponding attribute. This would be especially useful when `get_mask()` is called without any arguments.

The documentation should be expanded. For example, `get_mask()` doesn't have any documentation outside its docstring, but it really should.

mhvk commented:
> Based on the tests and the documentation the public API seems mostly reasonable. The one question I have is whether the output of `get_mask()` should be a structured array so that its components could be accessed by the name of the corresponding attribute. This would be especially useful when `get_mask()` is called without any arguments.

The way this is written, I think not - it is meant to give a array of plain bool back. It would not be too hard to extend (e.g., with a `structured=True/False` keyword argument), but I'm not sure I see a big benefit over just accessing the relevant masks directly.

mhvk commented on astropy/coordinates/baseframe.py:
```diff
@@ -359,6 +361,59 @@ def _insert(self, obj, values, axis=0):
 
+class CoordinateSharedMethods:
+    @property
+    def masked(self):
```
I still don't typically think of typing at all, but, yes, I can do that. If we keep the mixin, that actually would be a nice bit of self-contained annotated code (is it `ndarray[bool]` for `.mask`?)

eerovaher replied:
With [`numpy.typing.NDArray`](https://numpy.org/doc/2.0/reference/typing.html#numpy.typing.NDArray) the annotation can be simplified to `NDArray[np.bool]`.

mhvk replied:
I'm going to leave this to a typing PR; time is definitely running out.

eerovaher replied:
I am not allowed to demand that you write the annotations.

mhvk replied:
sorry, this really just a matter of spending time wisely - once a module is typed, I think requiring the rest to remain typed is fine, of course.

mhvk commented on astropy/coordinates/baseframe.py:
```diff
@@ -359,6 +361,59 @@ def _insert(self, obj, values, axis=0):
+        # Broadcast makes it readonly too.
+        return np.broadcast_to(combine_masks(masks), self.shape)
+
+    mask = property(
+        get_mask,
+        doc="""The mask associated with these coordinates.
+
+    Combines the masks of all components of the underlying representation,
+    including possible differentials.
+    """,
+    )
```
It would be an extra function call (as it would just contain `return self.get_mask()`).

eerovaher replied:
Mypyc should be able to optimize away such intermediate function calls, but we are still very far from compiling `coordinates`.

If you think this is code is that performance critical then the implementation makes sense, but do add a comment.

mhvk replied:
It is just standard use of `property` - it is certainly not performance critical.

nstarman replied:
Agreed. This happens in the Python docs https://docs.python.org/3/library/functions.html#property and is reasonable when we want a public `getter`.

mhvk commented on astropy/coordinates/representation/base.py:
```diff
@@ -426,6 +451,45 @@ def shape(self, shape):
+
+    def _ensure_masked(self):
+        """Ensure Masked components."""
+        # TODO: should we just allow the above property to be set?
```
I'll add to this note to state that it needs to remain consistent with `Time`.

mhvk replied:
Done

mhvk commented on astropy/coordinates/tests/test_masked.py:
```diff
@@ -0,0 +1,364 @@
+
+
+class MaskedSphericalSetup:
+    @classmethod
+    def setup_class(self):
```
In principle, yes, but we've used `self` often in the tests and it oddly reads better when one doesn't assign `cls.XXX`. But happy to change too; it will make the distinction with `setup_method()` clearer.

mhvk replied:
It is used all through astropy, wrongly or not. I don't think I should deviate from it here.

nstarman replied:
If we don't fix it here, we should fix it Astropy-wide.

neutrinoceros replied:
> It is used all through astropy, wrongly or not.

note that despite syntacticly colored, `self` and `cls` are not keywords in Python, merely conventions. So, if one uses `self` as the name of the first argument in a `@classmethod`, the `self` variable is *still* binded to the *class* (indeed, there's no way to access any specific instance from this argument alone), so, I'm really not sure what you mean by "wrongly or not", because I don't see how it could possibly be correct (and is certainly very confusing). I agree that we should fix it at the package level. Apparently this doesn't exist as a linting rule already so I'll start by opening a feature request on ruff and see if we can get detection (maybe even autofixing) for free.

mhvk replied:
Yes, let's auto-fix in a separate PR. Thanks!

eerovaher replied:
Another option would be to rewrite the tests to make use of module-level constants or `pytest` fixtures.

mhvk replied:
Since `ruff` can autofix this, I just went ahead and did it. I do find `setup_class` considerably clearer than fixtures... Sorry for gruffness...

mhvk commented on docs/whatsnew/7.0.rst:
```diff
@@ -288,6 +289,29 @@ New ``Lorentz2D`` model
+coordinate frames and coordinate representations.
+
+Like more generally for |Masked| instances, one can access the mask of using
+``.mask``, check whether the underlying data are masked using ``masked``,
```
I tried that, but it is not possible: `SkyCoord` does not show inherited members in the documentation. This is of course a general problem, in particular with the fact that `SkyCoord` uses `__getattr__` to have also all the attributes of the underlying frame. And I'd prefer not to "solve" this by referring to the `CoordinateMixin`...

eerovaher replied:
The main motivation of introducing `CoordinateSharedMethods` is to avoid code duplication, which is convenient for developers. However, if having the methods there hides their documentation from the users then that is inconvenient for them, and in public API user convenience is more important than developer convenience.

mhvk replied:
It turns out removing `CoordinateSharedMethods` means all 3 still need to be implemented in `SkyCoord`: `get_mask` because of the attributes, `mask` and `masked` because they are abstract properties that any actual class needs to implement. For now, I've kept this in a separate commit.

mhvk replied:
Done.

nstarman replied:
Why doesn't SkyCoord show inherited methods?

mhvk replied:
I actually do not know - but I remember trying it briefly and the docs build failed. Didn't quite have time to investigate further.

nstarman replied:
Ok. We should revisit this. Because I do think large classes like SkyCoord can benefit from being refactored to inherit from a set of **completely orthogonal** base classes.

nstarman commented:
In particular there is very little support for structured arrays outside of numpy.

mhvk commented:
I rebased and implemented most comments -- I ran out of time for additions to the documentation, but feel this is good enough for an experimental feature as it is.

Last ping also to @taldcroft - please check whether you like the basics behind adding masks to `SkyCoord` here. I will be focusing on Quantity 2.0 the coming months, so if it doesn't happen now, masked `SkyCoord` is unlikely to happen anytime soon.

eerovaher commented on docs/coordinates/index.rst:
```diff
@@ -425,6 +425,56 @@ and transforming velocities. These are available both via the lower-level
+
+In the last example, you will notice that the angles of the second item have
+become masked too. This is because the distance is used in the conversion
+(which involves matrix multiplications on the cartesian representation).
```
This is leaking implementation details to the user documentation. Using ICRS -> GCRS conversion would be a better example than ICRS -> FK5.

mhvk replied:
Ah, sorry, you had mentioned that before. I'll go ahead and change to GCRS.

eerovaher commented on astropy/coordinates/tests/test_masked.py:
```diff
@@ -0,0 +1,382 @@
+        assert_array_equal(self.sc.get_mask("ra", "dec"), False)
+
+    def test_filled(self):
+        unmasked = self.sc.unmasked
+        filled = self.sc.filled(unmasked[1])
```
The fill value here is a `SkyCoord` that has all the same components and frame attributes as the masked `SkyCoord`. Is it possible to use a fill value that doesn't specify some component? If not then that could be acceptable for now, but it should be documented.

mhvk replied:
See comment in main thread; this really is about `__setitem__` so I think out of scope here (especially as `.filled()` is defined in the `MaskableShapedLikeNDArray`).

eerovaher replied:
I use `SkyCoord.apply_space_motion()` quite a lot. When I have sources without radial velocity measurements it is reasonable to replace missing values with zeros, but if I have sources without proper motion measurements then assuming zero proper motion is usually not what I want at all. So the fact that it is not possible to fill only radial velocities without filling proper motions is important enough that it should be documented somewhere.

mhvk replied:
I am at a loss exactly what you are looking for here. As I mentioned, there is a big underlying problem, that we generally cannot set anything except for complete `SkyCoord`, which is not easy to solve at all. I don't see how I can document this usefully without documenting far more problems that currently exist, like the attributes mentioned below. 

More relevantly, this needs time I really do not have -- as usual before feature freeze, I have spent more time than I should have on reviewing lots of astropy stuff, and now I really should get to the referee report that is due tomorrow, to the two talks I'm supposed to give the coming weeks. If that means this PR does not get merged, so be it.

eerovaher reviewed (commented):
The example in the documentation using ICRS -> FK5 conversion should be rewritten. We are allowed to make documentation changes to v7.0 even after the feature freeze, but I'd prefer to make the change before if possible.

I am also starting to have some concerns about filling. The newly added tests check the filling works if the fill value includes all the components and frame attributes, but is that a requirement or is it possible to fill only a single component while leaving the others masked? For now it can be acceptable if filling only one component is not possible, but it sounds strange enough to me that it should be documented, preferably before this merged.

mhvk commented:
> I am also starting to have some concerns about filling. The newly added tests check the filling works if the fill value includes all the components and frame attributes, but is that a requirement or is it possible to fill only a single component while leaving the others masked?

Going back to `np.ma.MaskedArray`, the idea of `.filled()` is certainly that you get out something that is no longer masked. Whether that is the perfect solution, I don't know, but I do think the best way to find out is to have something for people to play with.

To be clear, the full implementation of `.filled()` is very simple, just get an unmasked copy and use `__setitem__`: https://github.com/astropy/astropy/blob/d7a6b978d1323e8bda860e50159819165b582a10/astropy/utils/masked/core.py#L412-L414

This also means that any problems it has are really problems for `__setitem__` -- and unfortunately `SkyCoord.__setitem__` has problems. For instance, the moment array-valued attributes are involved, it breaks easily:
```
sc = SkyCoord([0, 1], [2, 3], unit="deg", frame="altaz", pressure=[1000, 1010]*u.mbar)
sc[0] = sc[1]
# ValueError: can only set frame item from an equivalent frame
```
But I don't think it should be up to this PR to document and/or fix such problems; the idea here is to add an *experimental* feature, so the question should really be whether the basis is sound enough to allow experimentation. Plus of course that the basic idea is not a bad one, or too complex: I'm happy that you and @nstarman convinced me that my original special-casing of `Spherical` was a bad idea; the result is definitely better.

mhvk commented on astropy/coordinates/tests/helper.py:
```diff
@@ -11,7 +11,7 @@ def skycoord_equal(sc1, sc2):
-    eq = np.ones(shape=sc1.shape, dtype=bool)
+    eq = True
     for comp in sc1.data.components:
-        eq &= getattr(sc1.data, comp) == getattr(sc2.data, comp)
-    return np.all(eq)
+        eq &= np.all(getattr(sc1.data, comp) == getattr(sc2.data, comp))
+    return eq
```
The way it was set up, one started with an unmasked array of bool, but in-place `&=` on that with a masked array does not work (since the mask cannot be propagated to an unmasked-instance). So, this allows the code to work for masked components (properly ignoring elements that are masked).

nstarman reviewed (approved):
Just transferring my approval from the Representations PR. Contingent on other maintainer's' approval, of course.

mhvk commented on docs/whatsnew/7.0.rst:
```diff
@@ -420,6 +421,32 @@ simple example might be:
+`~astropy.coordinates.SkyCoord.mask` and check whether the underlying data are
+masked using `~astropy.coordinates.SkyCoord.masked`. Like for all masked
+classes, one can access underlying unmasked data using
+`~astropy.utils.masked.MaskableShapedLikeNDArray.unmasked` and one can get a
```
Sadly, because `SkyCoord` does not show inherited components, I can still not link directly to it for `unmasked` and `filled`.

mhvk commented on docs/coordinates/index.rst:
```diff
@@ -425,6 +425,60 @@ and transforming velocities. These are available both via the lower-level
+    <MaskedAngle [ 0.        , 15.00502838] deg>
+    >>> sc.separation_3d(sc[0])  # doctest: +FLOAT_CMP
+    <MaskedDistance [ 0., ———] kpc>
+    >>> from warnings import catch_warnings
```
The change to GCRS sadly means the `nan` give runtime warnings, which, with our default settings, cause test failures. If anyone knows a more elegant solution to have the tests pass without littering the documentation with this, please let me know... (in `io.fits`, tests are skipped for this reason, but don't really want to do that here).

eerovaher replied:
Because you mention the warnings in the text then it would be best to [declare them as part of the intended output and show them](https://github.com/scientific-python/pytest-doctestplus?tab=readme-ov-file#showing-warnings).

mhvk replied:
Thanks, that's super helpful!

mhvk commented:
OK, rebased and changelog entries combined.

eerovaher commented on docs/changes/coordinates/17016.feature.rst:
The change log entry seems too long and too detailed, especially given that this pull request has a What's New entry.

mhvk replied:
Yes, I guess it does duplicate. I'll shorten.

mhvk commented:
Docs now better, changelog entry shortened.

mhvk commented on docs/coordinates/index.rst:
```diff
@@ -425,6 +425,60 @@ and transforming velocities. These are available both via the lower-level
+    <MaskedAngle [ 0.        , 15.00502838] deg>
+    >>> sc.separation_3d(sc[0])  # doctest: +FLOAT_CMP
+    <MaskedDistance [ 0., ———] kpc>
+    >>> gcrs = sc.gcrs  # doctest: +SHOW_WARNINGS
```
So weird, no warnings emitted with "all dependencies".

mhvk replied:
I cannot reproduce this, maybe due to different architecture/compiler on MacOS. Will try `+IGNORE_OUTPUT`

mhvk replied:
Thankfully, that worked.

neutrinoceros commented:
@eerovaher I think (but haven't checked thoroughly) that your comments and suggestions were all addressed. Are you happy with the PR now ?

eerovaher commented:
I implemented `SkyCoord.partially_filled()` myself. With this addition, I would be willing to merge this pull request, but someone else should review the commit I made.

neutrinoceros commented:
on it

mhvk commented:
Separately, it looks like @eerovaher and I have reached an impasse. I think it would be best if one of the other coordinate maintainers could have a look and decide whether the PR as it stands is ready to go in. ping @adrn and @StuartLittlefair.

taldcroft commented:
@mhvk - thanks for this thorough PR, and sorry for being late to the review. This will finally make `SkyCoord` a fully-fledged mixin column, able to be used in table joins and other operations (though see next comment). 

I've read through the comments and it seems that essentially all of the review comments have been addressed. This new feature is clearly marked as new with a Warning admonition, so users will be aware that the code may be subject to some updates. Getting feedback from users will be helpful, and we can never forget that perfect is the enemy of good!

This is very important IMO and so I strongly urge that the coordinate maintainers approve this PR for inclusion in 7.0.

taldcroft commented:
Currently this masking feature is not enabled for use in table operations, but it just takes (I think) a one-line fix. With these changes, `SkyCoord` works in table operations and tests pass:
```
diff --git a/astropy/coordinates/baseframe.py b/astropy/coordinates/baseframe.py
index 35d28b1ded..986ce2ae58 100644
--- a/astropy/coordinates/baseframe.py
+++ b/astropy/coordinates/baseframe.py
@@ -148,6 +148,7 @@ class CoordinateFrameInfo(MixinInfo):
 
     attrs_from_parent = {"unit"}  # Unit is read-only
     _supports_indexing = False
+    mask_val = np.ma.masked
 
     @staticmethod
     def default_format(val):
diff --git a/astropy/table/tests/test_operations.py b/astropy/table/tests/test_operations.py
index 3ba1095d61..2ea85da0bb 100644
--- a/astropy/table/tests/test_operations.py
+++ b/astropy/table/tests/test_operations.py
@@ -661,7 +661,7 @@ class TestJoin:
 
         # Check for left, right, outer join which requires masking. Works for
         # the listed mixins classes.
-        if isinstance(col, (Quantity, Time, TimeDelta)):
+        if isinstance(col, (Quantity, Time, TimeDelta, SkyCoord)):
             out = table.join(t1, t2, join_type="left")
             assert len(out) == 3
             assert np.all(out["idx"] == [0, 1, 3])
@@ -1477,7 +1477,7 @@ class TestVStack:
         # Check for outer stack which requires masking.  Only Time supports
         # this currently.
         t2 = table.QTable([col], names=["b"])  # different from col name for t
-        if isinstance(col, (Time, TimeDelta, Quantity)):
+        if isinstance(col, (Time, TimeDelta, Quantity, SkyCoord)):
             out = table.vstack([t, t2], join_type="outer")
             assert len(out) == len_col * 2
             assert np.all(out["a"][:len_col] == col)
@@ -2075,7 +2075,7 @@ class TestHStack:
             assert np.all(out["col0_2"] == col2)
 
         # Time class supports masking, all other mixins do not
-        if isinstance(col1, (Time, TimeDelta, Quantity)):
+        if isinstance(col1, (Time, TimeDelta, Quantity, SkyCoord)):
             out = table.hstack([t1, t2], join_type="outer")
             assert len(out) == len(t1)
             assert np.all(out["col0_1"] == col1)
```

StuartLittlefair commented on docs/whatsnew/7.0.rst:
```diff
@@ -436,6 +437,32 @@ simple example might be:
+classes, one can access underlying unmasked data using
+`~astropy.utils.masked.MaskableShapedLikeNDArray.unmasked` and one can get a
+new instance with masked data replaced with a fill value using
+:meth:`~astropy.utils.masked.MaskableShapedLikeNDArray.filled`.
```
When I try and use this, the `filled` object is still masked. Is this intended? It's not what I expect. Using the example from the main docs....

```python
In [23]: gcrs.filled(gcrs2[1])
Out[23]: 
<SkyCoord (GCRS: obstime=J2000.000, obsgeoloc=(0., 0., 0.) m, obsgeovel=(0., 0., 0.) m / s): (ra, dec, distance) in (deg, deg, kpc)
    [(15.00054403, 2.99988395, 0.1), (        ———,        ———, ———)]>
```

```python
In [25]: gcrs.filled(gcrs2[1]).unmasked
Out[25]: 
<SkyCoord (GCRS: obstime=J2000.000, obsgeoloc=(0., 0., 0.) m, obsgeovel=(0., 0., 0.) m / s): (ra, dec, distance) in (deg, deg, kpc)
    [(15.00054403, 2.99988395, 0.1), (30.00201927, 3.99996188, 1. )]>
```

mhvk replied:
Thanks for noticing that! The issue arises because the fill value you are using is masked, and that the causes the output to become masked again as well (basically, `.filled()` just does `result = self.unmasked.copy(); result[self.mask] = fill_value`-- if you do `gcrs.filled(gcrs[1].unmasked)` it works properly.

Now this may explain it, but of course does not make it any less expected. What do you think would be best? Raising if `fill_value` is masked? Or a warning? Or ignoring the mask?

FWIW, for an actual array (or Quantity), the mask on the fill value is simply ignored, and I see that this holds for `Time` too -- so maybe I should ensure that happens here too... I'll look a bit later.

StuartLittlefair replied:
I’m relaxed about this remaining as it is. Once you’ve seen it, it’s easy for the user to workaround in the manner you suggest. 

If you have time we could ignore the mask on the fill value (I like consistency) but it should prevent merge in time for the big freeze

mhvk replied:
My last commit ensured the mask on the fill value is ignored; I agree it is best to be consistent, at least for now.

taldcroft commented:
@mhvk - thanks, the new code looks good, just waiting for tests to pass. In the meantime there is a conflict in the What's New file.

StuartLittlefair commented:
OK - I've found the time to look over this and generally speaking I like the implementation. 

I totally understand where @eerovaher is coming from in this instance. My primary use case for this would be the handling of transformations and space motion for coordinates with missing data (in my case, usually missing RVs in Gaia). I really do want to be able to do something like 

```python
rvs = Masked([0.1, np.nan]*u.km/u.s, mask=[False, True])
sc = SkyCoord(
    [1., 2.]*u.hourangle, [3., 4.]*u.deg, 
    distance=[0.1, 1.]*u.kpc, 
    pm_ra_cosdec=[100, 10]*u.mas/u.yr, 
    pm_dec=[-100, -10]*u.mas/u.yr, 
    radial_velocity=rvs
)
sc.fill_attrs({'radial_velocity': 0*u.km/u.s}).apply_space_motion(dt=10*u.yr)
```

I like the way the end result would still be masked appropriately, and I can get the unmasked result if I want. 
As it stands, using `filled` is really quite hard. OTOH I agree with @mhvk  that adding extra API to SkyCoord *in this PR* is the right idea.

I'm not familiar enough with the code in `utils.masked.core` to know if `fill_attrs` could easily be added there, but if so - and time allows - it would be an excellent addition to this PR. 

**BUT** I don't want to see that prevent this PR from being merged in time for feature freeze on this basis and I'm happy with everything else, so I will approve.

mhvk commented:
Let me take this out of the thread, since it is less about `.filled()` than about `__setitem__`: basically, as implemented, if one has an unmasked representation (or frame or `SkyCoord`) and uses `sc[...] = other_sc`, then if `other_sc` is masked, `sc` will become masked too. This in itself seems fairly logical, but it is not how it works for `ndarray` (where we have no choice: we cannot override `__setitem__`, and numpy just ignores the mask), nor for `Quantity` (which by itself does not know about masks, and cannot change its own class to `MaskedQuantity` anyway). 

For `Time`, we do check whether the value is masked, so it does not become masked either. So, options would seem to be:
1. Adjust the PR here to ignore the mask for `__setitem__` too;
2. Adjust `Time` to take the mask into account;
3. Live with the inconsistency.

I really do not like (3). Maybe the cleanest (as far as possible) is to go with (1) for now - I am traveling next 2 weeks but can probably do a quick bugfix for both `Time` and coordinates if we want to go with (2) instead.

mhvk commented:
OK, did the push, and tests pass. Very happy to see that table joining now works properly!

And, just to reiterate it, I'm very much in favour of following this up with adding other methods like `fill_attrs` or a `set_mask` or so.

taldcroft reviewed (approved):
:tada:

mhvk commented:
Thanks, @nstarman, @eerovaher, @taldcroft and @StuartLittlefair - it's going in!
